src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -206,28 +206,115 @@ internal object ReadODataOrCustomInstanceAnnotationValue(string annotationName,
+                            Debug.Assert(namespaceName != Metadata.EdmConstants.EdmNamespace, "If type was in the edm namespace it should already have been resolved");
+
+                            // if there is a payload type name that is not in the Edm namespace, assume complex
+                            // (it doesn't really matter; we just need someplace to hang the name)
+                            typeReference = new EdmComplexType(namespaceName, name, /*baseType*/ null, /*isAbstract*/true, /*isOpen*/ true).ToTypeReference(/*isNullable*/ true);
+                            return isCollection ? new EdmCollectionType(typeReference).ToTypeReference(/*isNullable*/ true) : typeReference;
+                        }
+
+                        typeReference = EdmCoreModel.Instance.GetString(/*isNullable*/ true);
+                    }
+                    else if (jsonReaderValue is bool)
+                    {
+                        typeReference = EdmCoreModel.Instance.GetBoolean(/*isNullable*/ true);
+                    }
+                    else if (jsonReaderValue is string)
+                    {
+                        typeReference = EdmCoreModel.Instance.GetString(/*isNullable*/ true);
+                    }
+                    else
+                    {
+                        typeReference = EdmCoreModel.Instance.GetDecimal(/*isNullable*/ true);
+                    }
+
+                    if (payloadTypeName != null)
+                    {
+                        TypeUtils.ParseQualifiedTypeName(payloadTypeName, out namespaceName, out name, out isCollection);
+                        typeReference = new EdmTypeDefinition(namespaceName, name, typeReference.PrimitiveKind()).ToTypeReference(/*isNullable*/ true);
+                    }
+
+                    return typeReference;
+
+                case JsonNodeType.StartObject:
+                    if (payloadTypeName != null && generateTypeIfMissing)
+                    {
+                        TypeUtils.ParseQualifiedTypeName(payloadTypeName, out namespaceName, out name, out isCollection);
+                        Debug.Assert(!isCollection, "NodeType is StartObject but payload specifies a collection.");
+                        return new EdmComplexType(namespaceName, name, /*baseType*/ null, /*isAbstract*/true, /*isOpen*/ true).ToTypeReference(/*isNullable*/ true);
+                    }
+
+                    return new EdmUntypedStructuredType().ToTypeReference(/*isNullable*/ true);
+
+                case JsonNodeType.StartArray:
+                    if (payloadTypeName != null && generateTypeIfMissing)
+                    {
+                        TypeUtils.ParseQualifiedTypeName(payloadTypeName, out namespaceName, out name, out isCollection);
+                        Debug.Assert(isCollection, "NodeType is start array but payload specifies a non-collection type");

```

Comment:
>Debug.Assert(isCollection, "NodeType is start array but payload specifies a non-collection type"); [](start = 24, length = 98)

same question. just wondering why this is a problem only in debug builds, but not in retail.